### DISCUSSION TITLE
Dictionary object of class with `nb::dynamic_attr`

### DISCUSSION THREAD

**Discussion by zhqrbitee** [wjakob/nanobind#275] [Q&A] [answered]
Hi, 

I have some legacy codes that use pybind11 and reply on operations on `__dict__` and I'm trying to move it to nanobind.  

A problem I hit is that the `__dict__` doesn't seems to be not created when using `nb::dynamic_attr`, so I'm curious on how `nb::dynamic_attr` works without `__dict__`?

I still can get the dictionary object in C++ codes, but it seems like a dirty hack:
```c++
#include "nanobind/nanobind.h"
#include "nanobind/stl/optional.h"

namespace nb = nanobind;
using namespace nb::literals;

struct A{};

static PyObject **nb_dict_ptr(PyObject *self) {
  PyTypeObject *tp = Py_TYPE(self);
  return (PyObject **) ((uint8_t *) self + tp->tp_dictoffset);
}

NB_MODULE(test_nanobind_bad_cast, m) {
  nb::class_<A>(m, "A", nb::dynamic_attr())
      .def("__init__", [](A *holder) {
        new (holder) A{};
      });

  m.def("_get_dict", [](nanobind::handle self) {
    return nanobind::handle(*nb_dict_ptr(self.ptr()));
  });
}
```

**Comment by wjakob** [comment, marked as answer]
IIRC the dictionaries are created on demand when you first write to them (this behavior may depend on the Python version)

**Reply by zhqrbitee** [reply]
Thanks for your answer, so it seems playing with this dictionary are dangerous and I may need to find other ways
